gtsam/hybrid/GaussianHybridFactorGraph.h

```diff
@@ -0,0 +1,181 @@
+using sharedFactor = boost::shared_ptr<Factor>;
+GTSAM_EXPORT std::pair<GaussianMixture::shared_ptr, sharedFactor>
+EliminateHybrid(const GaussianHybridFactorGraph& factors, const Ordering& keys);
+
+template <>
+struct EliminationTraits<GaussianHybridFactorGraph> {
+  typedef Factor FactorType;
+  typedef GaussianHybridFactorGraph FactorGraphType;
+  typedef GaussianMixture ConditionalType;
+  typedef HybridBayesNet BayesNetType;
+  typedef HybridEliminationTree EliminationTreeType;
+  typedef HybridBayesNet BayesTreeType;
+  typedef HybridEliminationTree JunctionTreeType;
+
+  /// The function type that does a single elimination step on a variable.
+  static std::pair<GaussianMixture::shared_ptr, sharedFactor> DefaultEliminate(
+      const GaussianHybridFactorGraph& factors, const Ordering& ordering) {
+    return EliminateHybrid(factors, ordering);
+  }
+};
+
+class GTSAM_EXPORT GaussianHybridFactorGraph
+    : public HybridFactorGraph<GaussianFactorGraph>,
+      public EliminateableFactorGraph<GaussianHybridFactorGraph> {
+ public:
+  using shared_ptr = boost::shared_ptr<GaussianHybridFactorGraph>;
+  using Base = HybridFactorGraph<GaussianFactorGraph>;
+
+ public:
+  /// Default constructor
+  GaussianHybridFactorGraph() = default;
+
+  /**
+   * @brief Construct a new Hybrid Factor Graph object.
+   *
+   * @param gaussianGraph A factor graph with continuous factors.
+   * @param discreteGraph A factor graph with only discrete factors.
+   * @param dcGraph A DCFactorGraph containing DCFactors.
+   */
+  GaussianHybridFactorGraph(const GaussianFactorGraph& gaussianGraph,
+                            const DiscreteFactorGraph& discreteGraph,
+                            const DCFactorGraph& dcGraph)
+      : Base(gaussianGraph, discreteGraph, dcGraph) {}
+
+  // Allow use of selected FactorGraph methods:
+  using Base::empty;
+  using Base::reserve;
+  using Base::size;
+  using Base::operator[];
+
+  /**
+   * Add a gaussian factor *pointer* to the internal gaussian factor graph
+   * @param gaussianFactor - boost::shared_ptr to the factor to add
+   */
+  template <typename FACTOR>
+  IsGaussian<FACTOR> push_gaussian(
+      const boost::shared_ptr<FACTOR>& gaussianFactor) {
+    factorGraph_.push_back(gaussianFactor);
+    Base::Base::push_back(gaussianFactor);
+  }
+
+  /// Construct a factor and add (shared pointer to it) to factor graph.
+  template <class FACTOR, class... Args>
+  IsGaussian<FACTOR> emplace_gaussian(Args&&... args) {
+    auto factor = boost::allocate_shared<FACTOR>(
+        Eigen::aligned_allocator<FACTOR>(), std::forward<Args>(args)...);
+    push_gaussian(factor);
+  }
+
+  /**
+   * @brief Add a single factor shared pointer to the hybrid factor graph.
+   * Dynamically handles the factor type and assigns it to the correct
+   * underlying container.
+   *
+   * @tparam FACTOR The factor type template
+   * @param sharedFactor The factor to add to this factor graph.
+   */
+  template <typename FACTOR>
```

Comment:
Why is this even a template? Factor is a common base class. But maybe that does not work with shared ptrs?

Reply:
We would need the template for SFINAE right? To make sure the push_back is happening to the correct factor graph?

Reply:
I don’t think this method is an SFINAE. Just dynamic casts

Reply:
Let me try it out. Also FYI we have a typedef above called `sharedFactor` which I will change to `SharedFactor` so that I can use it here.

Reply:
Also you're correct about it not being SFINAE. I was misunderstood what was really happening here and with a fresh mind (+ without the stress of getting something running for the state estimator) this makes a lot of sense.